Sources/FlatBuffers/FlatBuffer.swift

```diff
@@ -0,0 +1,222 @@
+    @discardableResult
+    func ensureSpace(size: UInt8) -> UInt8 {
+        if Int(size) + _writerSize > _capacity { reallocate(size) }
+        assert(size < FlatBufferMaxSize, FlatbufferError.growBeyondTwoGB.errorDescription ?? "FB doesn't support more than 2GB")
+        return size
+    }
+    
+    /// Reallocates the buffer incase the object to be written doesnt fit in the current buffer
+    /// - Parameter size: Size of the current object
+    fileprivate func reallocate(_ size: UInt8) {
+        let currentWritingIndex = writerIndex
+        while _capacity <= _writerSize + Int(size) {
+            _capacity = _capacity << 1
+        }
+        
+        /// solution take from Apple-NIO
+        _capacity = _capacity.convertToPowerofTwo
+        
+        let newData = UnsafeMutableRawPointer.allocate(byteCount: _capacity, alignment: alignment)
+        newData.initializeMemory(as: UInt8.self, repeating: 0, count: _capacity)
+        newData.advanced(by: writerIndex).copyMemory(from: _memory.advanced(by: currentWritingIndex), byteCount: _writerSize)
+        _memory.deallocate()
+        _memory = newData
+    }
+    
+    /// Clears the current size of the buffer
+    public func clearSize() {
+        _writerSize = 0
+    }
+    
+    /// Clears the current instance of the buffer, replacing it with new memory
+    public func clear() {
+        _writerSize = 0
+        alignment = 1
+        _memory.deallocate()
+        _memory = UnsafeMutableRawPointer.allocate(byteCount: _capacity, alignment: alignment)
+    }
+    
+    /// Resizes the buffer size
+    /// - Parameter size: new size for the buffer
+    internal func resize(_ size: Int) {
+        _writerSize = size
+    }
+    
+    /// Reads an object from the buffer
+    /// - Parameters:
+    ///   - def: Type of the object
+    ///   - position: the index of the object in the buffer
+    public func read<T>(def: T.Type, position: Int) -> T {
+        let size = MemoryLayout<T>.size
+        let r = UnsafeMutableRawPointer.allocate(byteCount: size, alignment: size)
+        r.copyMemory(from: _memory.advanced(by: position), byteCount: size)
+        return r.load(as: T.self)
+    }
+    
+    /// Reads a slice from the memory assuming a type of T
+    /// - Parameters:
+    ///   - index: index of the object to be read from the buffer
+    ///   - count: count of bytes in memory
+    public func readSlice<T>(index: Int32, count: Int32) -> [T] {
+        let array = UnsafeBufferPointer(start: _memory.advanced(by: Int(index)).assumingMemoryBound(to: T.self), count: Int(count))
+        return Array(array)
+    }
+    
+    /// Reads a string from the buffer and encodes it to a swift string
+    /// - Parameters:
+    ///   - index: index of the string in the buffer
+    ///   - count: length of the string
+    ///   - type: Encoding of the string
+    public func readString(at index: Int32, count: Int32, type: String.Encoding = .utf8) -> String? {
```

Comment:
@aardappel the `FlatBuffer` is actually `ByteBuffer` and since all the languages have their Read functions in there, I added it there too. I think if it made you confused we should rename it